### PR TITLE
SQL for device metadata replacement

### PR DESCRIPTION
**Pull request by ice-cronus**
ReplaceDeviceMetadata is used NoSQL for replace metadata in database, so DB checks (like if user exists) was never called.
This PR fixes that issue, ReplaceDeviceMetadata was rewritten using SQL to preserve DB references.

### CODE REVIEW

File: users/internal/device/metadata/metadata.go [hunk 1]
```diff
@@ -134,6 +136,103 @@ func (*repository) verifyDeviceAppVersion(metadata *DeviceMetadata) error {
...
+	sql := `REPLACE INTO DEVICE_METADATA 
+			(country_short,country_long,region,city,isp,latitude,longitude,domain,zipcode,timezone,net_speed,idd_code,area_code,weather_station_code,
+			weather_station_name,mcc,mnc,mobile_brand,elevation,usage_type,updated_at,first_install_time,last_update_time,user_id,device_unique_id,
+			readable_version,fingerprint,instance_id,hardware,product,device,type,tags,device_id,device_type,device_name,brand,carrier,manufacturer,
+			user_agent,system_name,system_version,base_os,build_id,bootloader,codename,installer_package_name,push_notification_token,api_level,
+			tablet,pin_or_fingerprint_set,emulator) 
+			VALUES                       
+			(:country_short,:country_long,:region,:city,:isp,:latitude,:longitude,:domain,:zipcode,:timezone,:net_speed,:idd_code,:area_code,:weather_station_code,
+			:weather_station_name,:mcc,:mnc,:mobile_brand,:elevation,:usage_type,:updated_at,:first_install_time,:last_update_time,:user_id,:device_unique_id,
+   			:readable_version,:fingerprint,:instance_id,:hardware,:product,:device,:type,:tags,:device_id,:device_type,:device_name,:brand,:carrier,:manufacturer,
```
Comment (ice-ares) [inline users/internal/device/metadata/metadata.go:150]: indent properly

File: users/internal/device/metadata/metadata.go [hunk 2]
```diff
@@ -104,18 +104,20 @@ func (r *repository) ReplaceDeviceMetadata(ctx context.Context, input *DeviceMet
...
 		return errors.Wrapf(err, "failed to get location information based on IP %v to replace device metadata", clientIP.String())
 	}
+	metadata.IP2Locationrecord = convertIP2Location(&ip2locationRecord)
 	before, err := r.GetDeviceMetadata(ctx, input.ID)
 	if err != nil && !errors.Is(err, storage.ErrNotFound) {
 		return errors.Wrapf(err, "failed to get current device metadata for %#v", input.ID)
 	}
-	var result []*Model
-	if err = r.db.ReplaceTyped("DEVICE_METADATA", metadata, &result); err != nil {
+	sql, params := r.prepareSQL(metadata)
```
Comment (ice-ares) [inline users/internal/device/metadata/metadata.go:116]: inline it

File: users/internal/device/metadata/contract.go [hunk 3]
```diff
@@ -98,9 +98,9 @@ var (
 type (
 	Model struct {
```
Comment (ice-ares) [inline users/internal/device/metadata/contract.go:99]: can this be private?

File: users/internal/device/metadata/metadata.go [hunk 4]
```diff
@@ -104,18 +104,20 @@ func (r *repository) ReplaceDeviceMetadata(ctx context.Context, input *DeviceMet
 	metadata := new(Model)
 	metadata.UpdatedAt = time.Now()
```
Comment (ice-ares) [inline users/internal/device/metadata/metadata.go:105]: move this inside DeviceMetadata, and make sure the field order is correct in DB

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: users/internal/device/metadata/contract.go (hunk 3 above)
Comment (ice-cronus) [reply users/internal/device/metadata/contract.go:99]: It seems no, it is used in fixture package in search for existing DB values ( AllExistingDeviceMetadata )

Comment (ice-ares) [reply users/internal/device/metadata/contract.go:103]: then move it in the public section

File: users/internal/device/metadata/metadata.go [hunk 5]
```diff
@@ -102,20 +103,19 @@ func (r *repository) ReplaceDeviceMetadata(ctx context.Context, input *DeviceMet
 		return vErr
 	}
 	metadata := new(Model)
-	metadata.UpdatedAt = time.Now()
 	metadata.DeviceMetadata = *input
+	metadata.UpdatedAt = time.Now()
```
Comment (ice-ares) [inline users/internal/device/metadata/metadata.go:107]: unnecessary change

Comment (ice-ares) [reply users/internal/device/metadata/metadata.go:107]: actually, you need to set updatedAt in the input

File: users/internal/device/metadata/metadata.go [hunk 6]
```diff
@@ -134,6 +134,174 @@ func (*repository) verifyDeviceAppVersion(metadata *DeviceMetadata) error {
 	return nil
 }
 
+//nolint:funlen // A lot of fields here.
+func (dm *Model) prepareSQL() (string, map[string]interface{}) {
```
Comment (ice-ares) [inline users/internal/device/metadata/metadata.go:138]: replaceSQL

**Review by ice-ares — APPROVED** [review]

**Pull request merged**